src/connectors/campaignService.ts

```diff
@@ -0,0 +1,209 @@
+import type {
+  Connections,
+  ValueOf,
+  CampaignStage,
+  Campaign,
+  User,
+} from 'definitions'
+
+import {
+  CAMPAIGN_TYPE,
+  CAMPAIGN_STATE,
+  CAMPAIGN_USER_STATE,
+} from 'common/enums'
+import {
+  ForbiddenByTargetStateError,
+  ForbiddenByStateError,
+  ForbiddenError,
+} from 'common/errors'
+import {
+  shortHash,
+  toDatetimeRangeString,
+  fromDatetimeRangeString,
+} from 'common/utils'
+import { AtomService } from 'connectors'
+
+interface Stage {
+  name: string
+  period?: readonly [Date, Date | undefined]
+}
+
+export class CampaignService {
+  private connections: Connections
+  private models: AtomService
+
+  public constructor(connections: Connections) {
+    this.connections = connections
+    this.models = new AtomService(connections)
+  }
+
+  public createWritingChallenge = async ({
+    name,
+    description,
+    link,
+    coverId,
+    applicationPeriod,
+    writingPeriod,
+    state,
+    creatorId,
+  }: {
+    name: string
+    description: string
+    link: string
+    coverId?: string
+    applicationPeriod: readonly [Date, Date]
+    writingPeriod: readonly [Date, Date]
+    state?: ValueOf<typeof CAMPAIGN_STATE>
+    creatorId: string
+  }) =>
+    this.models.create({
+      table: 'campaign',
+      data: {
+        shortHash: shortHash(),
+        type: CAMPAIGN_TYPE.writingChallenge,
+        name,
+        description,
+        link,
+        cover: coverId,
+        applicationPeriod: toDatetimeRangeString(
+          applicationPeriod[0],
+          applicationPeriod[1]
+        ),
+        writingPeriod: toDatetimeRangeString(
+          writingPeriod[0],
+          writingPeriod[1]
+        ),
+        state: state || CAMPAIGN_STATE.pending,
+        creatorId,
+      },
+    })
+
+  public updateStages = async (campaignId: string, stages: Stage[]) => {
```

Comment:
use `stages` as new stages would be more simple than update the diff?

Reply:
new stages will have new Ids and thus create new translation records ?

Reply:
yeah, and it's more easy to [return in order](https://github.com/thematters/matters-server/pull/4026/files#r1667990887)?